src/MSBuild/XMake.cs

```diff
@@ -2765,62 +2765,21 @@ DirectoryGetFiles getFiles
             // We need to look in a directory for a project file...
             if (projectFile == null)
             {
+                ValidateExtensions(projectsExtensionsToIgnore);
+                HashSet<string> extensionsToIgnore = new HashSet<string>(projectsExtensionsToIgnore ?? Array.Empty<string>(), StringComparer.OrdinalIgnoreCase);
                 // Get all files in the current directory that have a proj-like extension
-                string[] potentialProjectFiles = getFiles(projectDirectory ?? ".", "*.*proj");
+                string[] potentialProjectFiles = (getFiles(projectDirectory ?? ".", "*.*proj") ?? Array.Empty<string>()).Where(s =>
+                    !extensionsToIgnore.Contains(Path.GetExtension(s)) && !s.EndsWith("~", StringComparison.CurrentCultureIgnoreCase)).ToArray();
```

Comment:
I would rather avoid using LINQ here, since startup time is perf-sensitive and the manual loops are already written. Does this provide some benefit?

Reply:
I find this more straightforward and readable. It also involves adding fewer strings to the Dictionary/HashSet, which both reduces its size and reduces time looking through it. That being said, I can convert this back to for loops if you'd prefer with the same benefits. A little more verbose and possibly harder to read but still an improvement in my opinion.